Allow pickling of galois fields

fyellin opened the pull request:
Fix Bug #539. Allow pickling of galois fields by calling their static constructor.

fyellin committed:
Allow pickling of galois fields

mhostetter committed:
Dynamically create and test pickled objects

mhostetter commented:
I moved some code around. I also got rid of the once-created pickle binaries for testing. These can fail in the future, if a different `pickle` version is used or if the FieldArrayMeta internals change. Instead, each test run the pickles are generated and re-tested inside a new subprocess.

Let me know if you have more thoughts or suggestions, then I'll merge in.

mhostetter committed:
Add to changelog

fyellin commented:
Your changes are fine.  

I probably would have done things a little bit different, but it's your project not mine: 

I'm a little bit more hesitant than you are to test equality by comparing string representations. I would have probably compared the values by capturing the value `x.tolist()` and making sure the lists are the same. 

I'm wondering if there is an easy way to test uniqueness.  Unpickling a Galois field is not only supposed to return that field, it's supposed to return the already existing one if that field existed earlier. Maybe I'll add that later.

Is there a reason you added the pickle version? This should work for all Pickle versions.

mhostetter commented:
Did you close this by accident?

fyellin commented:
Oh. <expletive deleted>.

I misread the previous email and thought you had already merged it.  So I
deleted it on my side. I will redo it.

On Thu, Jan 1, 2026 at 3:27 PM mhostetter ***@***.***> wrote:

> *mhostetter* left a comment (mhostetter/galois#639)
> <https://github.com/mhostetter/galois/pull/639#issuecomment-3704227520>
>
> Did you close this by accident?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mhostetter/galois/pull/639#issuecomment-3704227520>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABC2LII32DNFF6UPDFIBBD34EWUMHAVCNFSM6AAAAACQNFSWASVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMBUGIZDONJSGA>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

mhostetter committed:
Make `FieldArray` subclasses hashable

mhostetter committed:
Improve unit tests

fyellin commented:
Sorry.  It's back.  Fortunately, github doesn't immediately purge deleted branches associated with PRs.

I noticed that there are pickle tests in fields/test_classes.py that happen to work because the GF already exists.  These tests pass by accident.  You may want to delete them.

mhostetter commented:
> Is there a reason you added the pickle version? This should work for all Pickle versions.

Unsure what you mean.

> I'm wondering if there is an easy way to test uniqueness. Unpickling a Galois field is not only supposed to return that field, it's supposed to return the already existing one if that field existed earlier. Maybe I'll add that later.

When `GF()` is called with the same parameters, it also returns a singleton `FieldArray` subclass. So it's impossible for their to be a different subclass that's "the same". I added an extra unit test though.

Also, I added a specific hash to `FieldArray` subclasses that is derived from the field properties.

I think this is probably good now. Let me know if you have other suggestions.

fyellin commented:
You used pickle.HIGHEST_PROTOCOL into your pickle.dump() call.  This change
should work across all pickle protocols. No big deal.

On Thu, Jan 1, 2026 at 4:18 PM mhostetter ***@***.***> wrote:

> *mhostetter* left a comment (mhostetter/galois#639)
> <https://github.com/mhostetter/galois/pull/639#issuecomment-3704261144>
>
> Is there a reason you added the pickle version? This should work for all
> Pickle versions.
>
> Unsure what you mean.
>
> I'm wondering if there is an easy way to test uniqueness. Unpickling a
> Galois field is not only supposed to return that field, it's supposed to
> return the already existing one if that field existed earlier. Maybe I'll
> add that later.
>
> When GF() is called with the same parameters, it also returns a singleton
> FieldArray subclass. So it's impossible for their to be a different
> subclass that's "the same". I added an extra unit test though.
>
> Also, I added a specific hash to FieldArray subclasses that is derived
> from the field properties.
>
> I think this is probably good now. Let me know if you have other
> suggestions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mhostetter/galois/pull/639#issuecomment-3704261144>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABC2LIJ25IJIRM4X2H5JGJT4EW2NLAVCNFSM6AAAAACQNFSWASVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMBUGI3DCMJUGQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

mhostetter committed:
Address @fyellin comments

mhostetter commented:
Ah, yes, thanks! I addressed your comments. I'll merge once CI passes unless there's anything else you'd like to add / see added.

mhostetter committed:
Merge branch 'develop' into fyellin/pickle
